### PR TITLE
Homography images

### PR DESCRIPTION
**Pull request by tomer-algo**
# Intro
  - Handle SmS homography images
# Project
  - Suggested Skeleton
# Implemented Sections
  - Generate Homography images for training (from SmS Selected frames)
  - Merge Datasets (add homography images to train / val / test datasets)
  - Add homography matrix to annotations (Save time in model analysis)
  - Did it require refactoring in other locations as well? No
# Tests Description
  - 
# Undeveloped Sections
  - 
# What's New
  - Generate homography images for training
  - more efficient analysis code 
# Checklist
  - [x] Module tested fully after last commit
  - [ ] Scheduled a PR meeting (if needed)
  - [ ] Documentation (comments or a dedicated documnet)
### Tested platforms
  - [x] Local
  - [ ] Cloud (Sagemaker)
  - [ ] Staging
  - [ ] Not relevant
### Tested Cameras
  - [ ] Gen 1.5
  - [ ] Gen 2
  - [x] Not relevant
# Comments
### Included Features
  - https://github.com/nanit/deep-high-resolution-net.pytorch/pull/18
  - https://github.com/nanit/deep-high-resolution-net.pytorch/pull/21
  - https://github.com/nanit/deep-high-resolution-net.pytorch/pull/22
### Infrastructure Dependencies
  - 
### Tech Comments
  - 
### Testing Instructions
  - 
### Knowledge
| / | Link |
| ------ | ------ |
| Monday | https://nanit.monday.com/boards/912051282/pulses/1429116871?userId=17446237 |
| TLD |  |
| Tech Spec |  |
### Todos
  -

### CODE REVIEW

File: smart_sheets/gen_homography_datasets.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+
+def homography_on_bbox(box_obj, homography_matrix):
+    points = np.array([[box_obj.x1, box_obj.y1],
+                       [box_obj.x2, box_obj.y1],
+                       [box_obj.x2, box_obj.y2],
+                       [box_obj.x1, box_obj.y2]]).reshape(4, 1, 2)
+    box_points_trans = cv.perspectiveTransform(points, homography_matrix)
+    x1_homography, y1_homography = np.min(box_points_trans, axis=0)[0]
+    x2_homography, y2_homography = np.max(box_points_trans, axis=0)[0]
+    box_homography_obj = BBox(x1=max(x1_homography, 0),
```
Comment (Nimrods) [inline smart_sheets/gen_homography_datasets.py:30]: What is the meaning of creating a bounding box if the homography is bad?
Meaning if the box is out of image boundaries, why clip it and not discard this image?

File: smart_sheets/gen_homography_datasets.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+                              x2=min(x2_homography, SKETCH_W),
+                              y2=min(y2_homography, SKETCH_H)
+                              )
+
+    return box_homography_obj
+
+
+def homography_on_point(point, homography_matrix):
+    p = np.array([[[float(point[0]), float(point[1])]]])
+    new_xy = cv.perspectiveTransform(p, homography_matrix)
```
Comment (Nimrods) [inline smart_sheets/gen_homography_datasets.py:41]: Where do you clip / check that the point is in the image boundaries?

File: smart_sheets/merge_annots.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+from smart_sheets.sms import homography
+
+ANNOT_FOLDER = os.path.expanduser('~/nanit/skeleton_data/skeleton_2021_10_28_db_data/annot')
+IMAGES_FOLDER = os.path.expanduser('~/nanit/deep-high-resolution-net.pytorch/data/nanit_mpii/images')
+HOMOGRAPHY_ANNOT_FOLDER = os.path.expanduser('~/nanit/skeleton_data_only_sms/data_after_homography/annot')
+OUTPUT_ANNOT_FOLDER = os.path.expanduser('~/nanit/skeleton_data/phase2_and_homography/annot')
+
+
+def main():
+    annot_paths = [os.path.join(ANNOT_FOLDER, f) for f in os.listdir(ANNOT_FOLDER)]
```
Comment (Nimrods) [inline smart_sheets/merge_annots.py:15]: sanity check that both lists have the same length?

File: smart_sheets/gen_homography_datasets.py (hunk 1 above)
Comment (tomer-algo) [reply smart_sheets/gen_homography_datasets.py:30]: that doesn't mean the homography is bad.
when doing homography, the bounding box is distorted (a bit) and then the vertices might fall off the image
it's related to how tight the bounding box, etc..

File: smart_sheets/gen_homography_datasets.py (hunk 2 above)
Comment (tomer-algo) [reply smart_sheets/gen_homography_datasets.py:41]: Will add it

File: smart_sheets/merge_annots.py (hunk 3 above)
Comment (tomer-algo) [reply smart_sheets/merge_annots.py:15]: will add it

File: smart_sheets/gen_homography_datasets.py (hunk 1 above)
Comment (Nimrods) [reply smart_sheets/gen_homography_datasets.py:30]: But I think we can expect the box will not cross image boundaries, right?
The baby shouldn't be at the edges (in SmS scenario)
it's not a wildlife scenraio...

Comment (tomer-algo) [reply smart_sheets/gen_homography_datasets.py:30]: I agree its a rare case,
but it might have some small negative value / value greater than image dim

Comment (Nimrods) [reply smart_sheets/gen_homography_datasets.py:30]: ok
if you think it's really not worth it, let's continue with it
But keep in mind that bad homography is worse then not doing homography at all :)

**Review by Nimrods — APPROVED** [review]

**Pull request merged**